README.md
# `trackerscraper`

`trackerscraper` is a tool for retrieving BitTorrent peer IP addresses from BitTorrent trackers.
It was quickly written by me for a university project.
As such it is very rough around the edges and most certainly not stable.
Read the disclaimer below.

## Features

Currently `trackerscraper` only supports [magnet links](http://bittorrent.org/beps/bep_0009.html)
(looks like `magnet:?xt=...`; often found on torrent index sites) and only supports
[UDP trackers](http://bittorrent.org/beps/bep_0015.html).
It features some rudimentary logging.
It outputs results and errors as JSON.
The error codes can be seen in `src/json.ml`.
It can handle escaped magnet links (`%3A` instead of `:` for example).

## Installation

### `opam` installation

Install using:
```
$ opam pin add trackerscraper git://github.com/Reperator/trackerscraper.git
```
`opam` should automatically install dependencies.

Uninstall using:
```
$ opam pin remove trackerscraper
```

### manual installation

Clone the repository and in the root directory run:
```
$ make
$ make install
```

You need to have the dependencies installed; one way to do this is:
```
$ opam install core lwt yojson jbuilder
```

Uninstall using:
```
$ make uninstall
```

## Usage

`trackerscraper` only has a command line interface.
Call it by typing `trackerscraper` into a shell, followed by a magnet link (in quotes) pointing to a torrent.
It will then attempt to fetch as many peer IP addresses as possible within the given constraints.
All optional parameters can be viewed by typing `trackerscraper -help`:

```
$ trackerscraper -help
query trackers for provided magnet link

  trackerscraper MAGNET-LINK

=== flags ===

  [--log logfile]     output logging information to logfile (-- for stdout)
  [--num-want n]      request n peers from a tracker (default: 80; < 255
                      required)
  [--output outfile]  output json result to outfile (-- for stdout)
                      (alias: -o)
  [--rescrape n]      request new peers n times (default: 0; may return the same
                      peers)
  [--timeout n]       wait for responses for n seconds
                      (alias: -t)
  [-pretty]           pretty-print json
  [-sequential]       scrape trackers sequentially instead of in parallel
  [-build-info]       print info about this build and exit
  [-version]          print the version of this build and exit
  [-help]             print this help text and exit
                      (alias: -?)
```

### Example output

This is some (censored) `trackerscraper` output. The JSON output format is not finalized and will most likely change.

```
$ trackerscraper --log logfile.txt --num-want 3 -pretty "magnet:?xt=..."
{
  "error": null,
  "errormsg": null,
  "hash": "deadbeefdeadbeefdeadbeefdeadbeefdeadbeef",
  "name": "example-torrent",
  "peers": [],
  "numberTrackers": 1,
  "trackers": [
    {
      "leechers": 123,
      "seeders": 456,
      "peers": [
        { "ip": "4.3.2.1", "port": 11111 },
        { "ip": "8.7.6.5", "port": 22222 },
        { "ip": "1.2.2.1", "port": 33333 }
      ],
      "numberPeers": 3,
      "error": null,
      "errormsg": null,
      "hostname": "tracker.example.org",
      "ip": "1.2.3.4",
      "port": 1337
    }
  ]
}
$ cat logfile.txt
[2017-10-29 02:20:16.703198+02:00] parsed magnet link "magnet:?xt=..." with result:
info_hash: deadbeefdeadbeefdeadbeefdeadbeefdeadbeef
name: example-torrent
udp trackers:
  tracker.example.org/1.2.3.4:1337
http trackers:
  tracker.example.org/1.2.3.4:1337/some/path
peers:

[2017-10-29 02:20:16.706529+02:00] sent connect (16 bytes, tid: beefcace) to 1.2.3.4:1337
[2017-10-29 02:20:16.733152+02:00] received connect (16 bytes, connection_id: beefdeadbeefdead) from 1.2.3.4:1337
[2017-10-29 02:20:16.733209+02:00] sent announce started (98 bytes, cid: beefdeadbeefdead, tid: bedeadef, num_want 3) to 1.2.3.4:1337
[2017-10-29 02:20:16.758171+02:00] received announce response (80 bytes, cid: beefdeadbeefdead, 12 leechers, 809 seeders, 3 peers) from 1.2.3.4:1337
[2017-10-29 02:20:16.758190+02:00] attempting to fetch 821 peers from 1.2.3.4:1337
[2017-10-29 02:20:16.758215+02:00] sent announce stopped (98 bytes, tid feedface) to 1.2.3.4:1337
```

### Suggested parameters

- `--num-want 200`: During testing trackers returned up to 200 peers per announce request (may be due to path MTU).

- `-t 2` or `-t .5`: Non-responsive trackers prevent `trackerscraper` from finishing. Reduce timeout to below the default 15s.

- `--log --`: Enable logging to stdout to see live information.

- `--rescrape 10`: Request more peers 10 times to get massively more peers than with a single request.

## Internals

### Language

`trackerscraper` is written in OCaml.
`trackerscraper` requires OCaml 4.05.0 or newer.
`trackerscraper` is built using [Jbuilder](https://github.com/janestreet/jbuilder).

### Dependencies

`trackerscraper` uses [core](https://github.com/janestreet/core), [Lwt](https://github.com/ocsigen/lwt) and
[Yojson](https://github.com/mjambon/yojson).

### Documentation

There is none (yet).
`trackerscraper -help` prints all command line options.
This README serves as documentation for the time being.

### Modules

#### `Net`

Client-side Implementation of the BitTorrent UDP Tracker Protocol.

#### `Magnet` and `Magnet_lex`

Magnet link parsing. `Magnet_lex` is generated by `ocamllex`.
`Magnet` may at some point be replaced by a `menhir` parser.

#### `Util`

Utility functions such as hex-encode/decode.

#### `Json`

Json backend.

#### `Trackerscraper`

Main module; contains command line interface and setup code.

### Behavior

`trackerscraper` tries to mask as a [Transmission](https://transmissionbt.com/) 2.92 client to prevent detection by trackers.
Its default values are similar and it generates its `peer_id` the same way Transmission does.
After it is done fetching addresses it disconnects from the tracker to prevent its host's IP address
from being listed as a peer; this does not work reliably (on the tracker's side).

`trackerscraper` uses Lwt to contact multiple trackers in parallel (this behavior can be turned off with `-sequential`).

`trackerscraper` is fairly performant; during testing I was able to fetch around 4000 IP addresses from two trackers
within one second with a round-trip time of ~40ms.

### Testing

There are no tests (yet).

## Future development

### Planned

- Lots of refactoring; currently the code is somewhat chaotic

- Support for http trackers

- Support for bencoded .torrent files

- Revamping the command line interface to allow for more flexibility

- Revamping the JSON backend


### Uncertain

- Support for [DHT](http://bittorrent.org/beps/bep_0005.html)

- Support for [PEX](http://bittorrent.org/beps/bep_0011.html)

- Backends other than JSON

## License

`trackerscraper` is licensed under the LGPL 3.0. See `LICENSE`.

## Disclaimer

I have no idea about the legal aspects of using this tool.

Using it most likely violates the TOS of many torrent sites.

Contacting trackers may add your IP address to the pool of peers.

Other peers may see your IP address and attempt to connect.

IP addresses can be located (assuming no precautions were taken).

Use at your own risk.


CHANGES.md
# Changes

## unreleased changes

## 0.3.2

- Fallback to stdout if log/output file could not be created (due to permissions/non-existent 
directories).
- OCaml 4.04.0 no longer supported due to changes in string/bytes handling.

## 0.3.1

- Fixed escaped magnet link handling.

## 0.3.0

- First public release.

src/net.ml
open Core
open Lwt
open Lwt_unix
open Printf

type connection = {
  hash : string;
  socket : Lwt_unix.file_descr;
  mutable cid : [`Active of string | `Inactive | `Closed]; (* connection_id *)
  local_port : int;
  peer_id : string; (* the peer_id we used *)
  address : string (* human readable address for logging (IPv4:Port) *)
}

type scrape_result = {
  leechers : int;
  seeders : int;
  peers : (Unix.Inet_addr.t * int) list
}

type action =
  | Connect
  | Announce
  | Scrape
  | Err

exception Tracker_error of string

exception Invalid_response of string *
  [ `Too_short
  | `Wrong_tid
  | `Invalid_action of int
  | `Unexpected_action of action]

exception Internal_error of string

let get_address { address } = address

let timeout = ref 15.

let set_timeout f = timeout := f

let action_to_string a =
  match a with
  | Connect -> "connect"
  | Announce -> "announce"
  | Scrape -> "scrape"
  | Err -> "error"

let action_of_int i =
  match i with
  | 0 -> Connect
  | 1 -> Announce
  | 2 -> Scrape
  | 3 -> Err
  | _ -> raise @@ Invalid_argument "unknown action"

let log_exn ~log address exn =
  let log_msg =
    match exn with
    | Tracker_error msg ->
      sprintf "tracker error: \"%s\" from %s" msg address
    | Invalid_response (response, reason) ->
      let reason =
        match reason with
        | `Too_short -> "response too short"
        | `Wrong_tid -> "transaction_id not matching"
        | `Invalid_action i -> sprintf "invalid action %d" i
        | `Unexpected_action a -> sprintf "unexpected action %s" (action_to_string a) in
      sprintf "invalid response (%s) from %s\n" reason address
    | Timeout ->
      sprintf "received no response within %fs from %s" !timeout address
    | Internal_error msg ->
      sprintf "internal error (%s)" msg
    | exn ->
      sprintf "other exception (%s)" (Exn.to_string exn) in
  log log_msg

let handle_tracker_error response len tid =
  match String.sub response ~pos:0 ~len:4 with
  | "\x00\x00\x00\x03" ->
    if String.sub response ~pos:4 ~len:4 <> tid then
      fail @@ Invalid_response (response, `Wrong_tid)
    else
      fail @@ Tracker_error (String.sub response ~pos:8 ~len:(len - 8))
  | "\x00\x00\x00\x00"
  | "\x00\x00\x00\x01"
  | "\x00\x00\x00\x02" as a ->
    let action = action_of_int (Util.int_of_4bytes a ~pos:0) in
    fail @@ Invalid_response (response, `Unexpected_action action)
  | _ ->
    fail @@ Invalid_response (response, `Invalid_action (Util.int_of_4bytes response ~pos:0))

let prepare_connect tid =
  (* magic number, action = connect, random transaction_id*)
  ("\x00\x00\x04\x17\x27\x10\x19\x80\x00\x00\x00\x00" ^ tid)
  |> Bytes.unsafe_of_string_promise_no_mutation (* only passed to write syscall *)

let verify_connect response len tid =
  let response = Bytes.unsafe_to_string response in
  if len < 16 then
    fail @@ Invalid_response (response, `Too_short)
  else if String.sub response ~pos:0 ~len:4 <> "\x00\x00\x00\x00" then
    handle_tracker_error response len tid
  else if String.sub response ~pos:4 ~len:4 <> tid then
    fail @@ Invalid_response (response, `Wrong_tid)
  else
    return @@ String.sub response ~pos:8 ~len:8 (* return the connection_id the tracker provided *)

let connect ~log ~hash ~addr ~port =
  let readable_address = sprintf "%s:%d" (Unix.Inet_addr.to_string addr) port in
  try%lwt
    let socket = socket PF_INET SOCK_DGRAM 0 in
    connect socket (ADDR_INET (addr, port)) >>= fun () ->

    let local_port =
      match getsockname socket with (* TODO this is not caught by the try%lwt *)
      | ADDR_UNIX _ -> raise @@ Internal_error "invalid socket"
      | ADDR_INET (_, port) -> port in

    let tid = Util.random_string 4 in
    let msg = prepare_connect tid in
    let%lwt send_len = write socket msg 0 (Bytes.length msg) in
    log @@ sprintf "sent connect (%d bytes, tid: %s) to %s"
      send_len
      (Util.to_hex tid)
      readable_address;

    let recv_buf = Bytes.create 10_000 in
    let%lwt recv_len =
      with_timeout !timeout (fun () -> read socket recv_buf 0 (Bytes.length recv_buf)) in
    let%lwt cid = verify_connect recv_buf recv_len tid in
    log @@ sprintf "received connect (%d bytes, connection_id: %s) from %s"
      recv_len
      (Util.to_hex cid)
      readable_address;

    let connection = {
      hash;
      socket;
      cid = `Active cid;
      local_port;
      peer_id = Util.peer_id ();
      address = readable_address
    } in
    (* apparently we do not need to keep a pointer to the invalidating thread *)
    ignore (sleep 60. >|= fun () -> connection.cid <- `Inactive);
    return connection
  with
  | e ->
    log_exn ~log readable_address e;
    fail e

let reconnect ~log ({ socket; address; _ } as connection) =
  try%lwt
    let tid = Util.random_string 4 in
    let msg = prepare_connect tid in
    let%lwt send_len = write socket msg 0 (Bytes.length msg) in
    log @@ sprintf "sent reconnect (%d bytes, tid: %s) to %s"
      send_len
      (Util.to_hex tid)
      address;

    let recv_buf = Bytes.create 10_000 in
    let%lwt recv_len =
      with_timeout !timeout (fun () -> read socket recv_buf 0 (Bytes.length recv_buf)) in
    let%lwt cid = verify_connect recv_buf recv_len tid in
    log @@ sprintf "received reconnect (%d bytes, connection_id: %s) from %s"
      recv_len
      (Util.to_hex cid)
      address;

    connection.cid <- `Active cid;

    ignore (sleep 60. >|= fun () -> connection.cid <- `Inactive);
    return_unit
  with
  | e ->
    log_exn ~log address e;
    fail e

let prepare_announce cid tid info_hash peer_id local_port num_want event =
  let event =
    match event with
    | `None -> '\x00'
    | `Completed -> '\x01'
    | `Started -> '\x02'
    | `Stopped -> '\x03' in

  (* constants are action = announce, downloaded, left, uploaded, IP address = 0 *)
  sprintf
    "%s\
     \x00\x00\x00\x01\
     %s\
     %s\
     %s\
     \x00\x00\x00\x00\x00\x00\x00\x00\
     \xff\xff\xff\xff\xff\xff\xff\xff\
     \x00\x00\x00\x00\x00\x00\x00\x00\
     \x00\x00\x00%c\
     \x00\x00\x00\x00\
     %s\
     \x00\x00\x00%c\
     %c%c"
    cid (* connection_id *)
    tid (* transaction_id *)
    (Util.from_hex info_hash)
    peer_id
    event
    (Util.random_string 4) (* we'll just generate 4 random bytes for the key *)
    (num_want land 0xff |> Char.of_int_exn)
    ((local_port lsr 8) land 0xff |> Char.of_int_exn) (* upper and lower byte of the port *)
    (local_port land 0xff |> Char.of_int_exn)
  |> Bytes.unsafe_of_string_promise_no_mutation (* only passed to write syscall *)

let verify_announce response len tid =
  let response = Bytes.unsafe_to_string response in
  if len < 20 then
    fail @@ Invalid_response (response, `Too_short)
  else if String.sub response ~pos:0 ~len:4 <> "\x00\x00\x00\x01" then
    handle_tracker_error response len tid
  else if String.sub response ~pos:4 ~len:4 <> tid then
    fail @@ Invalid_response (response, `Wrong_tid)
  else
    let leechers = Util.int_of_4bytes response ~pos:12 in
    let seeders = Util.int_of_4bytes response ~pos:16 in

    let num_peers = (len - 20) / 6 in
    let rec loop acc n =
      if n >= 0 then
        let offset = (20 + (6 * n)) in
        let addr = (* Inet_addr.t is represented as a string internally, this is very unsafe *)
          (Obj.magic (String.sub response ~pos:offset ~len:4) : Unix.Inet_addr.t) in
        let port =
          (Char.to_int response.[offset + 4] lsl 8) + (Char.to_int response.[offset + 5]) in
        loop ((addr, port) :: acc) (n - 1)
      else acc in

    return { leechers; seeders; peers = loop [] (num_peers - 1) }

let rec announce ~log ({ hash; socket; cid; local_port; peer_id; address } as c) ~num_want event =
  match cid with
  | `Closed ->
    log @@ sprintf "connection already closed (%s)" address;
    fail @@ Internal_error "connection already closed"
  | `Inactive ->
    log @@ sprintf "connection timed out, reconnecting (%s)" address;
    reconnect ~log c >>= fun () ->
    announce ~log c ~num_want event
  | `Active cid ->
    try%lwt
      let tid = Util.random_string 4 in
      let msg = prepare_announce cid tid hash peer_id local_port num_want event in
      let%lwt send_len = write socket msg 0 (Bytes.length msg) in
      log @@ sprintf "sent announce %s (%d bytes, cid: %s, tid: %s, num_want %d) to %s"
        (match event with `Started -> "started" | `None -> "none")
        send_len
        (Util.to_hex cid)
        (Util.to_hex tid)
        num_want
        address;

      let recv_buf = Bytes.create 10_000 in
      let%lwt recv_len =
        with_timeout !timeout (fun () -> read socket recv_buf 0 (Bytes.length recv_buf)) in
      let%lwt { leechers; seeders; peers } as result = verify_announce recv_buf recv_len tid in
      log @@ sprintf
        "received announce response (%d bytes, cid: %s, %d leechers, %d seeders, %d peers) from %s"
        recv_len
        (Util.to_hex cid)
        leechers
        seeders
        (List.length peers)
        address;

      return result
    with
    | e ->
      log_exn ~log address e;
      fail e

let close ~log ({ hash; socket; cid; local_port; peer_id; address } as c) =
  match cid with
  | `Closed ->
    log @@ sprintf "connection already closed (%s)" address;
    return_unit
  | `Inactive ->
    log @@ sprintf "connection already inactive (%s)" address;
    c.cid <- `Closed;
    close socket
  | `Active cid ->
    let tid = Util.random_string 4 in
    let msg = prepare_announce cid tid hash peer_id local_port 0 `Stopped in
    let%lwt send_len = write socket msg 0 (Bytes.length msg) in
    log @@ sprintf "sent announce stopped (%d bytes, tid %s) to %s"
      send_len
      (Util.to_hex tid)
      address;
    c.cid <- `Closed;
    close socket


src/json.ml
open Core
open Yojson.Safe
open Printf

type ok_result = {
  info_hash : string;
  name : string option;
  peers : (string option * Unix.Inet_addr.t * int) list;
  udp_trackers : (string option * Unix.Inet_addr.t * int * (Net.scrape_result, exn) result) list
}

let ip_str = Unix.Inet_addr.to_string

let default d = function
  | None -> d
  | Some v -> v

let handle_peers =
  List.map
    ~f:(fun (host, ip, port) -> `Assoc [
        ("hostname", match host with None -> `Null | Some name -> `String name);
        ("ip", `String (ip_str ip));
        ("port", `Int port)
      ])

let handle_trackers =
  List.map
    ~f:(fun (host, ip, port, result) ->
      let open Net in
      let error, errormsg =
        match result with
        | Error (Tracker_error msg) ->
          `Int 1, `String ("tracker error: " ^ msg)
        | Error (Invalid_response (_, `Too_short)) ->
          `Int 2, `String "tracker response too short"
        | Error (Invalid_response (_, `Wrong_tid)) ->
          `Int 3, `String "tracker response has wrong tid"
        | Error (Invalid_response (_, `Invalid_action a)) ->
          `Int 4, `String (sprintf "tracker response has invalid action code %d" a)
        | Error (Invalid_response (_, `Unexpected_action a)) ->
          `Int 5, `String (sprintf "tracker response has unexpected action %s" (action_to_string a))
        | Error (Internal_error msg) ->
          `Int 6, `String ("internal error: " ^ msg)
        | Error (Lwt_unix.Timeout) ->
          `Int 7, `String "request timed out"
        | Error exn ->
          `Int 0, `String ("unknown error: " ^ Exn.to_string exn)
        | Ok _ ->
          `Null, `Null in

      let default_fields = [
        ("error", error);
        ("errormsg", errormsg);
        ("hostname", match host with None -> `Null | Some name -> `String name);
        ("ip", `String (ip_str ip));
        ("port", `Int port)
      ] in

      match result with
      | Error _ ->
        `Assoc default_fields
      | Ok { leechers; seeders; peers } ->
        let peers =
          List.map
            peers
            ~f:(fun (addr, port) ->
              `Assoc [
                ("ip", `String (ip_str addr));
                ("port", `Int port)
              ]) in
        `Assoc (("leechers", `Int leechers)
        :: ("seeders", `Int seeders)
        :: ("peers", `List peers)
        :: ("numberPeers", `Int (List.length peers))
        :: default_fields))

let to_json result =
  match result with
  | Error (Magnet.Broken_link `No_info_hash) ->
    `Assoc [
      ("error", `Int 1);
      ("errormsg", `String "broken link: no info hash in magnet link")
    ]
  | Error (Magnet.Broken_link `Other) ->
    `Assoc [
      ("error", `Int 2);
      ("errormsg", `String "broken link: other lexer failure")
    ]
  | Error exn ->
    `Assoc [
      ("error", `Int 0);
      ("errormsg", `String ("unkown error: " ^ Exn.to_string exn))
    ]
  | Ok { info_hash; name; peers; udp_trackers } ->
    `Assoc [
      ("error", `Null);
      ("errormsg", `Null);
      ("hash", `String info_hash);
      ("name", match name with None -> `Null | Some n -> `String n);
      ("peers", `List (handle_peers peers));
      ("numberTrackers", `Int (List.length udp_trackers));
      ("trackers", `List (handle_trackers udp_trackers))
    ]


src/util.ml
open Core

let () = Random.self_init ()

type log = string -> unit

let ci = Char.to_int
let ic = Char.of_int_exn

let offset c =
  match c with
  | '0' .. '9' -> 48
  | 'a' .. 'f' -> 87
  | 'A' .. 'F' -> 55
  | _ -> raise @@ Invalid_argument "invalid hex digit"

let from_hex str =
  let str = (* if str is not of even length we pad with a '0' *)
    if String.length str mod 2 <> 0 then "0" ^ str else str in

  let hex_to_int c =
    ci c - offset c in

  String.init
    (String.length str / 2)
    ~f:(fun i -> ic ((hex_to_int str.[2 * i] lsl 4) + (hex_to_int str.[(2 * i) + 1])))

let to_hex ?(upper = false) str =
  let f i =
    let c = ci str.[i / 2] in
    let nibble = if i mod 2 = 0 then c lsr 4 else c land 0xf in
    let offset =
      if nibble < 0xa then 48 (* 0x0 <= nibble <= 0xf *)
      else if upper then 55 else 87 in
    ic (nibble + offset) in

  String.init (String.length str * 2) ~f

let random_string len =
  String.init len ~f:(fun _ -> Random.int 256 |> ic)

let peer_id () = (* faking a Transmission 2.92 peer id *)
  let pool = "0123456789abcdefghijklmnopqrstuvwxyz" in
  let pool_len = String.length pool in
  "-TR2920-" ^ (String.init 12 ~f:(fun _ -> pool.[Random.int pool_len]))

let int_of_4bytes str ~pos =
  (ci str.[pos] lsl 24)
  + (ci str.[pos + 1] lsl 16)
  + (ci str.[pos + 2] lsl 8)
  + (ci str.[pos + 3])

let unescape str =
  let len = String.length str in
  let out = Buffer.create len in (* out is <= str, so we won't have to reallocate *)
  let rec loop in_index =
    if in_index < len then
      if str.[in_index] = '%' then begin
        let new_char =
          try
            let first = str.[in_index + 1] in
            let second = str.[in_index + 2] in
            ((ci first - offset first) lsl 4) + (ci second - offset second)
          with
          | Invalid_argument _ (* index out of bounds *) ->
            raise @@ Invalid_argument "% too close to end of string" in
        ic new_char
        |> Buffer.add_char out;
        loop (in_index + 3)
      end else begin
        Buffer.add_char out str.[in_index];
        loop (in_index + 1)
      end in
  loop 0;
  Bytes.unsafe_to_string (Buffer.to_bytes out) (* obviously safe, no need to copy *)


src/magnet.ml
open Core
open Lwt

type magnet_info = {
  info_hash : string;
  name : string option;
  udp_trackers : (string option * Unix.Inet_addr.t * int) list;
  http_trackers : (string option * Unix.Inet_addr.t * int * string) list;
  peers : (string option * Unix.Inet_addr.t * int) list
}

exception Broken_link of [`No_info_hash | `Other]

let parse_magnet_link ~log link =
  let open Magnet_lex in
  let lexbuf = Lexing.from_string link in
  let%lwt info_hash =
    try%lwt return @@ prefix lexbuf with
    | Failure _ ->
      log @@ sprintf "no info_hash found in magnet link \"%s\" (possibly non-standard position)"
        link;
      fail @@ Broken_link `No_info_hash in

  let empty_info = {
    info_hash;
    name = None;
    udp_trackers = [];
    http_trackers = [];
    peers = []
  } in

  let rec loop info =
    let addr_of_str addr =
      let open Unix.Inet_addr in
      try None, Some (of_string addr) with
      | _ ->
        try Some addr, Some (of_string_or_getbyname addr) with
        | _ -> None, None in

    try%lwt match read lexbuf with
    | Name name -> loop { info with name = Some name }
    | Udp_tracker (addr, port) ->
      let new_info =
        match addr_of_str addr with
        | _, None -> info
        | None, Some addr ->
          { info with udp_trackers = (None, addr, port) :: info.udp_trackers }
        | Some hostname, Some addr ->
          { info with udp_trackers = (Some hostname, addr, port) :: info.udp_trackers } in
      loop new_info
    | Http_tracker (addr, port, path) ->
      let new_info =
        match addr_of_str addr with
        | _, None -> info
        | None, Some addr ->
          { info with http_trackers = (None, addr, port, path) :: info.http_trackers }
        | Some hostname, Some addr ->
          { info with http_trackers = (Some hostname, addr, port, path) :: info.http_trackers } in
      loop new_info
    | Peer (addr, port) ->
      let new_info =
        match addr_of_str addr with
        | _, None -> info
        | None, Some addr ->
          { info with peers = (None, addr, port) :: info.peers }
        | Some hostname, Some addr ->
          { info with peers = (Some hostname, addr, port) :: info.peers } in
      loop new_info
    | EOF -> return info
    with
    | Failure _ as exn ->
      log @@ sprintf "lexer failure: %s" (Exn.to_string exn);
      fail @@ Broken_link `Other in

  let%lwt result = loop empty_info in
  let summarise_udp_trackers_and_peers (host, ip, port) =
    sprintf "  %s%s:%d\n"
      (match host with Some name -> name ^ "/" | None -> "")
      (Unix.Inet_addr.to_string ip)
      port
  and summarise_http_trackers (host, ip, port, path) =
    sprintf "  %s%s:%d%s\n"
      (match host with Some name -> name ^ "/" | None -> "")
      (Unix.Inet_addr.to_string ip)
      port
      path in
  log @@ sprintf
    "parsed magnet link \"%s\" with result:\ninfo_hash: %s\nname: %s\n\
     udp trackers:\n%shttp trackers:\n%speers:\n%s"
    link
    result.info_hash
    (match result.name with Some str -> str | None -> "")
    (List.fold
      result.udp_trackers
      ~init:""
      ~f:(fun acc tracker -> acc ^ summarise_udp_trackers_and_peers tracker))
    (List.fold
      result.http_trackers
      ~init:""
      ~f:(fun acc tracker -> acc ^ summarise_http_trackers tracker))
    (List.fold
      result.peers
      ~init:""
      ~f:(fun acc peer -> acc ^ summarise_udp_trackers_and_peers peer));

  return @@ result


src/trackerscraper.ml
open Core
open Lwt
open Printf
open Out_channel
open Command

let trackerscraper_version = "0.3.2"

let summary = "query trackers for provided magnet link"

let run magnet log out pretty num_want rescrape seq timeout =
  let log =
    match log with
    | None -> ignore
    | Some file ->
      let log_fun oc str =
        fprintf oc "[%s] %s\n" Time.(to_string (now ())) str;
        flush oc in
      let oc =
        if file = "--" then stdout
        else
          try create file with
          | Sys_error _ (* permisson denied or invalid path *) ->
            log_fun stdout (sprintf "could not create log file %s, falling back to stdout\n" file);
            stdout in
      log_fun oc in

  let oc =
    match out with
    | None | Some "--" -> stdout
    | Some file ->
      try create file with
      | Sys_error _ (* permisson denied or invalid path *) ->
        log (sprintf "could not create output file %s, falling back to stdout\n" file);
        stdout in

  let output_json =
    if pretty then
      Yojson.Safe.pretty_to_channel oc
    else
      Yojson.Safe.to_channel oc in

  begin match timeout with
  | Some time -> Net.set_timeout time
  | None -> ()
  end;

  try%lwt
    let%lwt Magnet.{ info_hash; name; udp_trackers; peers; http_trackers } =
      Magnet.parse_magnet_link ~log magnet in

    let scrape_tracker (host, addr, port) =
      try%lwt
        let%lwt connection = Net.connect ~log ~hash:info_hash ~addr ~port in

        let announce = Net.announce ~log connection in

        let%lwt Net.{ leechers; seeders; peers } as first_scrape = announce ~num_want `Started in

        let required = leechers + seeders in
        log @@ sprintf "attempting to fetch %d peers from %s" required (Net.get_address connection);
        let rescrapes = ref rescrape in
        let peer_set =
          Hash_set.Poly.of_list first_scrape.Net.peers in

        while%lwt Hash_set.length peer_set < required && !rescrapes > 0 do
          let%lwt Net.{ peers; _ } =
            announce ~num_want:(Int.min num_want (required - Hash_set.length peer_set)) `None in
          List.iter peers ~f:(Hash_set.add peer_set);
          decr rescrapes;
          return_unit
        done >>= fun () ->
        Net.close ~log connection >>= fun () ->
        let scrape_result = Net.{ first_scrape with peers = Hash_set.to_list peer_set } in
        return (host, addr, port, Ok scrape_result)
      with
      | exn -> return (host, addr, port, Error exn) in

    let%lwt scrape_results =
      Lwt_list.(if seq then map_s else map_p) scrape_tracker udp_trackers in
    let json =
      Json.to_json @@ Ok Json.{ info_hash; name; peers; udp_trackers = scrape_results } in
    output_json json;
    return_unit
  with
  | exn ->
    let json = Json.to_json @@ Error exn in
    output_json json;
    return_unit

let command =
  basic_spec ~summary ?readme:None
    Spec.(
      empty
      +> anon ("magnet-link" %: string)
      +> flag "--log" (optional string)
        ~doc:"logfile output logging information to logfile (-- for stdout)"
      +> flag "--output" (optional string) ~aliases:["-o"]
        ~doc:"outfile output json result to outfile (-- for stdout)"
      +> flag "-pretty" no_arg
        ~doc:" pretty-print json"
      +> flag "--num-want" (optional_with_default 80 int)
        ~doc:"n request n peers from a tracker (default: 80; < 255 required)"
      +> flag "--rescrape" (optional_with_default 0 int)
        ~doc:"n request new peers n times (default: 0; may return the same peers)"
      +> flag "-sequential" no_arg
        ~doc:" scrape trackers sequentially instead of in parallel"
      +> flag "--timeout" (optional float) ~aliases:["-t"]
        ~doc:"n wait for responses for n seconds"
    ) (fun magnet log out pretty num_want rescrape seq timeout () ->
        Lwt_main.run (run magnet log out pretty num_want rescrape seq timeout))

let () =
  let build_info =
    sprintf
      "trackerscraper version: %s\n\
       OCaml version: %s\n\
       Execution mode: %s\n\
       OS type: %s\n\
       Executable: %s"
      trackerscraper_version
      Sys.ocaml_version
      (match Sys.execution_mode () with `Native -> "native" | `Bytecode -> "bytecode")
      Sys.os_type
      Sys.executable_name in
  Command.run ~version:trackerscraper_version ~build_info command
